azworker: disable time-sync

RaduBerinde opened the pull request:
Hyper-V time sync is causing backward time jumps (probably because
NTP is also enabled); adding a script to disable it.

RaduBerinde committed:
azworker: disable time-sync

Hyper-V time sync is causing backward time jumps (probably because
NTP is also enabled); adding the script to disable it to the worker VM.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16939)
<!-- Reviewable:end -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[build/bootstrap/disable-hyperv-timesync.sh, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KoUJM3-3RIZyfoeH4QY:-KoUJM3-3RIZyfoeH4QZ:b-djpqx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d329f5bde48460e0241c154abcbaad02e70b52b/build/bootstrap/disable-hyperv-timesync.sh#L28)):*
> ```Shell
> echo -n "${dev_id}" | sudo tee "${UNBIND_PATH}" > /dev/null
> if [ "$?" -ne "0" ]; then
>   echo "Time sync device already unbound, or error encountered."
> ```

You don't want to fail here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KoUJVHMQ9Ge2g_0ea6P:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[build/bootstrap/disable-hyperv-timesync.sh, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KoUJM3-3RIZyfoeH4QY:-KoUKnekBIiiVwqY_eGI:bux23sn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d329f5bde48460e0241c154abcbaad02e70b52b/build/bootstrap/disable-hyperv-timesync.sh#L28)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You don't want to fail here?
</blockquote></details>

It would fail `azworker create`.. I figured if it's already disabled for whatever reason, it should all be good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[build/bootstrap/disable-hyperv-timesync.sh, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KoUJM3-3RIZyfoeH4QY:-KoULpRVnTMVHpMygr1x:bydmiat) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d329f5bde48460e0241c154abcbaad02e70b52b/build/bootstrap/disable-hyperv-timesync.sh#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It would fail `azworker create`.. I figured if it's already disabled for whatever reason, it should all be good.
</blockquote></details>

Worried more about the "error encountered" case, but I'm OK not to sink too much time into this one. Would just be shame if it rotted and someone else would have to find out the hard way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[build/bootstrap/disable-hyperv-timesync.sh, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KoUJM3-3RIZyfoeH4QY:-KoUgqulcrXP4x0QFHZa:b-9f1ic1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d329f5bde48460e0241c154abcbaad02e70b52b/build/bootstrap/disable-hyperv-timesync.sh#L28)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Worried more about the "error encountered" case, but I'm OK not to sink too much time into this one. Would just be shame if it rotted and someone else would have to find out the hard way.
</blockquote></details>

Ok, changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Thanks!

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KoUkS2zoC5ZX9Ibq8H7:b-ucfknh)*
<!-- Sent from Reviewable.io -->

tamird commented:
Is this new script different from `pkg/acceptance/terraform/azure/disable-hyperv`? can we just have one, not both?

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-Kogs4j4K0nelbtfNIuk:b-7h3lju)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Hm, I didn't know about that one. Looks the same, except that one also enables `ntp`. AFAICT the azworker system already has NTP sync enabled via `systemd-timesyncd`, not sure what happens if you also install `ntp`.

RaduBerinde commented:
Suggestions how I can reuse that one? I don't want to mess with the terraform stuff and have to test it..

tamird commented:
I would just extract a script that does the shared part and modify the terraform script to call it - perhaps @cuongdo can suggest a test that isn't too onerous.

RaduBerinde committed:
azworker: support multiple workers

RaduBerinde commented:
I reused the same script and I added a commit that makes the script support multiple workers. I have to test a few cases, and I'm not sure how to test the terraform change (trivial as it is).

tamird commented:
r3 commit message still says "adding the script to disable it." r4 commit message body can just as soon be omitted.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 5 of 5 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[scripts/azworker.sh, line 55 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KohPJ59zCEvs1RfheSu:-KohPJ59zCEvs1RfheSv:b-jrv03p) ([raw file](https://github.com/cockroachdb/cockroach/blob/b48b4f1e6e29f28ab5a5bc9313c8c799490c5363/scripts/azworker.sh#L55)):*
> ```Shell
>     rsync -az "$(dirname "${0}")/../build/bootstrap/" "${USER}@${FQDN}:bootstrap/"
>     rsync -az "$(dirname "${0}")/../build/parallelbuilds-"* "${USER}@${FQDN}:bootstrap/"
>     rsync -az "$(dirname "${0}")/../build/disable-hyperv-timesync.sh" "${USER}@${FQDN}:./"
> ```

why does this one need a different directory? i'd put it in bootstrap, too.

---

*[scripts/azworker.sh, line 27 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KohPUxfxkJr3jm4S-pW:-KohPUxfxkJr3jm4S-pX:b32vuwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/76c33e5b206ced759e703eb19e5f09cbd1551880/scripts/azworker.sh#L27)):*
> ```Shell
> case ${1-} in
>     create)
>     if azure group show "${RG}" >/dev/null; then
> ```

nit: can you check that this passes shellcheck?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KohP6vfMLrWOIP-ZEtY:b6ifuw9)*
<!-- Sent from Reviewable.io -->

benesch commented on scripts/azworker.sh:
```diff
@@ -72,7 +76,12 @@ case ${1-} in
     delete)
-    azure group delete "${RG}"
+    azure vm delete "${RG}" "${NAME}"
+    if azure group show "${RG}" | grep "Type *: *virtualMachines" >/dev/null; then
```
`grep -q` instead of `grep ... > /dev/null`? Don't go out of your way.

RaduBerinde replied:
`azure` doesn't like when the pipe gets closed early

benesch reviewed (commented):
"azworker: support multiple workers" looks awesome, thanks @RaduBerinde!

RaduBerinde commented:
The script errors out with:
`Failed to stop ntp.service: Unit ntp.service not loaded.`

I am having trouble believing that this actually works in other cases on Azure. Should I just remove this section from the script?

RaduBerinde commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[scripts/azworker.sh, line 55 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KohPJ59zCEvs1RfheSu:-KohRQvJKWLoBvFVnS_i:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b48b4f1e6e29f28ab5a5bc9313c8c799490c5363/scripts/azworker.sh#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why does this one need a different directory? i'd put it in bootstrap, too.
</blockquote></details>

Fixed.

---

*[scripts/azworker.sh, line 27 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KohPUxfxkJr3jm4S-pW:-KohRPDHiaqFsexkd04f:b-nxdbq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/76c33e5b206ced759e703eb19e5f09cbd1551880/scripts/azworker.sh#L27)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: can you check that this passes shellcheck?
</blockquote></details>

Done (it does).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939)*
<!-- Sent from Reviewable.io -->

tamird commented:
We should probably ignore the exit code of `sudo service ntp stop`

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[scripts/azworker.sh, line 55 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KohPJ59zCEvs1RfheSu:-KohRvYnljkrmJo0FxaG:b-fk9w6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b48b4f1e6e29f28ab5a5bc9313c8c799490c5363/scripts/azworker.sh#L55)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Fixed.
</blockquote></details>

Doesn't look like it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KohSAsuor_FDDXSoUny:bjqblg2)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Well we get the same error for `start` so the whole thing seems useless..

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KohSwlvU8JKtZA_SxmI:bbtdddk)*
<!-- Sent from Reviewable.io -->

tamird commented:
Oh, well don't we want to run ntp? It seems like we're just not doing it right, but we should still do it.

I wonder if that stuff works in terraform for some reason? I'd suggest commenting it out with a TODO and filing an issue for @cuongdo 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KohTNvGFDra9M5WEpCI:b7mnbtb)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[scripts/azworker.sh, line 55 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-KohPJ59zCEvs1RfheSu:-KohU6BVnNUi2c8klcZk:b-blj8hc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b48b4f1e6e29f28ab5a5bc9313c8c799490c5363/scripts/azworker.sh#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Doesn't look like it
</blockquote></details>

Pushed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
The system already has ntp enabled (not via the `ntp` service, but via `systemd-timesyncd`), at least the azworker one.

In fact, I don't think the HyperV timesync would have been a problem in the first place if NTP wasn't also enabled - I believe the backward jumps are because they're stepping on each other's toes.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KohUAJCujnUJbhoQTl0:bejysjt)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
I'm testing these changes with Terraform now. Also, I vaguely recall that when the HyperV timesync was enabled but ntp wasn't, we got clock-related errors. However, that could be a faulty recollection on my part.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KohfpgKoO9CWvS3ADcF:br4rm9c)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Tested with one of the nightlies and verified that the time sync service is still disabled correctly

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-KohisreWg0vQwaoFhsb:burlo6k)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks @cuongdo - the problem is that the change as is does not work with azworker because `sudo service ntp stop/start` in the script fails. My question is: do the lines below actually do something useful under Terraform? My hunch is that they error out the same (and the error is ignored) so we can just remove them.

```
# Install ntpdate.
sudo apt-get -qqy install ntpdate
# Stop ntp, force ntp sync, restart ntp.
sudo service ntp stop
sudo ntpdate -b ntp.ubuntu.com
sudo service ntp start
```

> Also, I vaguely recall that when the HyperV timesync was enabled but ntp wasn't, we got clock-related errors

Ubuntu 16.04 has time synchronization enabled by default (not via the `ntp` service, but through `systemd-timesyncd`) [1](https://www.digitalocean.com/community/tutorials/how-to-set-up-time-synchronization-on-ubuntu-16-04)
I'm not sure if this is the same OS used with Terraform?

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-Koi4ey2rKro9DPHHI6D:b-9w9jtv)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
scripts: update ntp commands in disable-hyperv-timesync.sh

The commands to set up NTP fail on Ubuntu 16.04 or later (causing the azworker
script to fail). Ubuntu 16.04 uses `systemd-timesyncd` instead of `ntp` by
default. Updating the commands accordingly.

RaduBerinde commented:
Added a commit that removes those lines.. This is what I'd like to merge if it doesn't change anything on terraform.

cuongdo commented:
Yes, our Terraform-based nightlies use Ubuntu 16.04.

I'll test the changes. I'm wary of removing `ntpdate`, because VMs sometimes start up with clocks that need quite a bit of syncing, but we'll see what the tests yield.

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-Koi727m11YPCn3cj0vg:b-4l61jp)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Why are you weary of removing `ntpdate`? It was replaced by `timedatectl`, which is set up by default.. We could add a `timedatectl set-ntp true` just for good measure.

RaduBerinde commented:
Updated script, we need to remove the nic and public-ip when we have multiple workers.

RaduBerinde commented:
Ah, you were talking about the `ntpdate -b` command, which I now see forces a NTP sync. I'm not sure how to force a sync with Ubuntu 16.04's default time sync stuff..

RaduBerinde commented:
Sorry for all the back and forth. I think I have the right thing now, which is to still force a sync using `ntpdate`, but update the commands to disable and re-enable NTP using Ubuntu's new mechanism.

tamird commented:
Reviewed 1 of 5 files at r3, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
@cuongdo ok to merge this? (I put the `ntpdate` thing back in). Given that it's the same OS I don't think we need to retest the script under terraform specifically.

cuongdo commented:
Should be fine. Thanks for tackling this.

On Tue, Jul 11, 2017 at 9:24 AM RaduBerinde <notifications@github.com>
wrote:

> @cuongdo <https://github.com/cuongdo> ok to merge this? (I put the ntpdate
> thing back in). Given that it's the same OS I don't think we need to retest
> the script under terraform specifically.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/16939#issuecomment-314442031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABffplAEqvm_RS8S5hUoWgORY77Ag0T_ks5sM3dqgaJpZM4ORhub>
> .
>

RaduBerinde commented:
Thanks Cuong.

I had to make one more fix to the azworker script - apparently a VM can still show up in the list right after a `vm delete`, so we are now counting the number of VMs before deletion.

tamird commented:
Reviewed 1 of 5 files at r3, 2 of 2 files at r9, 1 of 1 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[scripts/azworker.sh, line 83 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-Kom5a6-LMWzhcGrMGqZ:-Kom5a6-LMWzhcGrMGq_:b-d8o6e4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bc7d67f0f89f2ce77e31aca9df11a65bae946d0/scripts/azworker.sh#L83)):*
> ```Shell
>     # deleted VM doesn't disappear right away from the listing. So we instead
>     # count the initial number of VMs.
>     NUM_VMS=$(azure group show "${RG}" | (grep "Type *: *virtualMachines" || true) | wc -l)
> ```

`grep -c`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[scripts/azworker.sh, line 83 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-Kom5a6-LMWzhcGrMGqZ:-Kom6sK8XC5QM3humxGX:buqr79r) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bc7d67f0f89f2ce77e31aca9df11a65bae946d0/scripts/azworker.sh#L83)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`grep -c`?
</blockquote></details>

Silly reason to make another change and test again if you ask me, but done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> again

---

Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[scripts/azworker.sh, line 83 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-Kom5a6-LMWzhcGrMGqZ:-Kom86w1xI_JMJGvsC_1:b-h86n6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bc7d67f0f89f2ce77e31aca9df11a65bae946d0/scripts/azworker.sh#L83)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Silly reason to make another change and test again if you ask me, but done.
</blockquote></details>

A sillier reason would be changing these double quotes to single quotes!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16939#-:-Kom8AfN_3mz53136aJT:b-865gvj)*
<!-- Sent from Reviewable.io -->
